x/finality/keeper/msg_server.go

```diff
@@ -333,3 +333,24 @@ func (ms msgServer) validateActivationHeight(ctx sdk.Context, height uint64) (ui
 	}
 	return activationHeight, nil
 }
+
+// EquivocationEvidence handles the evidence of equivocation message sent from the finality gadget cw contract
+func (ms msgServer) EquivocationEvidence(goCtx context.Context, req *types.MsgEquivocationEvidence) (*types.MsgEquivocationEvidenceResponse, error) {
+	ctx := sdk.UnwrapSDKContext(goCtx)
+	evidence := req.Evidence
```

Comment:
After we have all fields of `Evidence` in `req`, we can validate it, construct the `Evidence` object, and then do the following logic

Reply:
> we can validate it

[evidence.IsSlashable()](https://github.com/babylonlabs-io/babylon/blob/base/consumer-chain-support/x/finality/types/finality.go#L73-L100) is enough, right?

Reply:
since we will not be using `Evidence`, we can't use this function:
<img width="623" alt="image" src="https://github.com/user-attachments/assets/9431297a-1137-47c1-92e9-a2d58977242c">

@SebastianElvis do you mean we should copy the logic in `ValidateBasic` here?

Reply:
Yep, in theory `MsgEquivocationEvidence` can be anything and users might submit invalid `MsgEquivocationEvidence`. It needs to go through the same validation process like this.